### PR TITLE
Added imvotenet image branch pretrain

### PR DESCRIPTION
**Pull request by THU17cyz**

### CODE REVIEW

File: configs/imvotenet/imvotenet_img_pretrain_4x2_sunrgbd-3d-10class.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+            type='CrossEntropyLoss', reduction='sum', loss_weight=1.0),
+        size_res_loss=dict(
+            type='SmoothL1Loss', reduction='sum', loss_weight=10.0 / 3.0),
+        semantic_loss=dict(
+            type='CrossEntropyLoss', reduction='sum', loss_weight=1.0)),
+)
+
+# model training and testing settings
+train_cfg = dict(
+    rpn=dict(
```
Comment (ZwwWayne) [inline configs/imvotenet/imvotenet_img_pretrain_4x2_sunrgbd-3d-10class.py:137]: We may use img_rpn, img_rpn_proposal, and img_rcnn to distinguish them from potential similar configs for votenet.

File: configs/imvotenet/imvotenet_img_pretrain_4x2_sunrgbd-3d-10class.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+log_config = dict(
+    interval=50,
+    hooks=[
+        dict(type='TextLoggerHook'),
+        dict(type='TensorboardLoggerHook')
+    ])
+# yapf:enable
+dist_params = dict(backend='nccl')
+log_level = 'INFO'
+load_from = 'data/ckpts/mask_rcnn_r50_caffe_fpn_mstrain-\
```
Comment (ZwwWayne) [inline configs/imvotenet/imvotenet_img_pretrain_4x2_sunrgbd-3d-10class.py:291]: We can simply use the network address `http://download.openmmlab.com/mmdetection/v2.0/mask_rcnn/mask_rcnn_r50_caffe_fpn_mstrain-poly_3x_coco/mask_rcnn_r50_caffe_fpn_mstrain-poly_3x_coco_bbox_mAP-0.408__segm_mAP-0.37_20200504_163245-42aa3d00.pth`

File: configs/imvotenet/imvotenet_img_pretrain_4x2_sunrgbd-3d-10class.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        dir_class_loss=dict(
+            type='CrossEntropyLoss', reduction='sum', loss_weight=1.0),
+        dir_res_loss=dict(
+            type='SmoothL1Loss', reduction='sum', loss_weight=10.0),
+        size_class_loss=dict(
+            type='CrossEntropyLoss', reduction='sum', loss_weight=1.0),
+        size_res_loss=dict(
+            type='SmoothL1Loss', reduction='sum', loss_weight=10.0 / 3.0),
+        semantic_loss=dict(
+            type='CrossEntropyLoss', reduction='sum', loss_weight=1.0)),
```
Comment (ZwwWayne) [inline configs/imvotenet/imvotenet_img_pretrain_4x2_sunrgbd-3d-10class.py:132]: comma is unnecessary.

File: configs/imvotenet/imvotenet_img_pretrain_4x2_sunrgbd-3d-10class.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+            neg_pos_ub=-1,
+            add_gt_as_proposals=True),
+        pos_weight=-1,
+        debug=False),
+    pos_distance_thr=0.3,
+    neg_distance_thr=0.6,
+    sample_mod='vote')
+
+test_cfg = dict(
+    rpn=dict(
```
Comment (ZwwWayne) [inline configs/imvotenet/imvotenet_img_pretrain_4x2_sunrgbd-3d-10class.py:182]: img_rpn, img_rcnn

File: configs/imvotenet/imvotenet_img_pretrain_4x2_sunrgbd-3d-10class.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+        max_num=1000,
+        nms_thr=0.7,
+        min_bbox_size=0),
+    rcnn=dict(
+        score_thr=0.05,
+        nms=dict(type='nms', iou_threshold=0.5),
+        max_per_img=100),
+    # soft-nms is also supported for rcnn testing
+    # e.g., nms=dict(type='soft_nms', iou_threshold=0.5, min_score=0.05)
+    sample_mod='seed',
```
Comment (ZwwWayne) [inline configs/imvotenet/imvotenet_img_pretrain_4x2_sunrgbd-3d-10class.py:195]: Put votenet related test config into `pts=dict()`.

**Review by ZwwWayne — APPROVED** [review]

File: mmdet3d/models/dense_heads/imvote_head.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+# from mmdet3d.core.post_processing import aligned_3d_nms
+# from mmdet3d.models.builder import build_loss
+# from mmdet3d.models.losses import chamfer_distance
+# from mmdet3d.models.model_utils import VoteModule
+# from mmdet3d.ops import build_sa_module, furthest_point_sample
+# from mmdet.core import build_bbox_coder, multi_apply
+
+# from .base_conv_bbox_head import BaseConvBboxHead
+
```
Comment (Tai-Wang) [inline mmdet3d/models/dense_heads/imvote_head.py:19]: Commented import could be removed?

File: mmdet3d/models/detectors/imvotenet.py [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+from mmdet.models import DETECTORS, build_backbone, build_head, build_neck
+from .base import Base3DDetector
+
+
+@DETECTORS.register_module()
+class ImVoteNet(Base3DDetector):
+    """ImVoteNet model.
+
+    https://arxiv.org/pdf/2001.10692.pdf
+    """
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/imvotenet.py:12]: r"""`ImVoteNet <https://arxiv.org/abs/2001.10692>`_ for 3D detection."""
It seems there exists the same problem in the docstring of VoteNet. Please also modify that accordingly.

File: mmdet3d/models/detectors/imvotenet.py [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+            gt_bboxes (list[Tensor]): Ground truth bboxes for each image with
+                shape (num_gts, 4) in [tl_x, tl_y, br_x, br_y] format.
+            gt_labels (list[Tensor]): class indices corresponding to each box
+            gt_bboxes_ignore (None | list[Tensor]): specify which bounding
+                boxes can be ignored when computing the loss.
+            gt_masks (None | Tensor) : true segmentation masks for each box
+                used if the architecture supports a segmentation task.
+            proposals : override rpn proposals with custom proposals. Use when
+                `with_rpn` is False.
+        Returns:
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/imvotenet.py:202]: Need a blank line before Returns.

File: mmdet3d/models/detectors/imvotenet.py [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+                `mmdet/datasets/pipelines/formatting.py:Collect`.
+            gt_bboxes (list[Tensor]): Ground truth bboxes for each image with
+                shape (num_gts, 4) in [tl_x, tl_y, br_x, br_y] format.
+            gt_labels (list[Tensor]): class indices corresponding to each box
+            gt_bboxes_ignore (None | list[Tensor]): specify which bounding
+                boxes can be ignored when computing the loss.
+            gt_masks (None | Tensor) : true segmentation masks for each box
+                used if the architecture supports a segmentation task.
+            proposals : override rpn proposals with custom proposals. Use when
+                `with_rpn` is False.
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/imvotenet.py:201]: Need a blank line before Returns.

File: mmdet3d/models/detectors/imvotenet.py [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+                      points=None,
+                      img=None,
+                      img_metas=None,
+                      gt_bboxes=None,
+                      gt_labels=None,
+                      gt_bboxes_ignore=None,
+                      gt_masks=None,
+                      proposals=None,
+                      **kwargs):
+        """
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/imvotenet.py:184]: Better to have one-sentence summary.

**Review by Tai-Wang — APPROVED** [review]

**Comment by hellock** [comment]
Task linked: [CU-5dqv4g ImVoteNet Faster R-CNN Pre-training (#217)](https://app.clickup.com/t/5dqv4g)

**Pull request merged**